### PR TITLE
fix(Flow cell Retrieval) Fix search pattern to find all flow cells

### PR DESCRIPTION
**Pull request by Vince-janv**
## Description
See issue for problem description: https://github.com/Clinical-Genomics/cg/issues/2583

This PR adds the three encryption directories we've used in the past. This allows us to iterate trough them and be sure that each archived flow cell will be covered by a search pattern.

All functional changes are in `cg.meta.backup.backup.BackupAPI.query_pdc_for_flow_cell`. The remaining changes are to allow the directories to get properly passed from the config to the function and improving tests.

### Added

- Two more encryption directories
- Iteration trough encryption directories when searching for archived flow cells

### Fixed

- Test is now a bit better

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b fix-fc-search-pattern -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: tests/meta/backup/conftest.py [hunk 1]
```diff
@@ -1,7 +1,27 @@
...
+
+    def mock_method(search_pattern: str) -> list[str]:
+        match = fnmatch.filter(archived_flow_cells, search_pattern)
+        if not match:
+            raise subprocess.CalledProcessError(
+                cmd="dummy_method", returncode=PDCExitCodes.NO_FILES_FOUND
+            )
+        return match
+
+    return mock_method
```
Comment (Vince-janv) [inline tests/meta/backup/conftest.py:23]: I went a bit wild here. Let me know if it is too unclear 😅

File: cg/meta/backup/backup.py [hunk 2]
```diff
@@ -253,20 +254,24 @@ def _set_flow_cell_status_to_retrieved(self, flow_cell: Flowcell):
...
-        except subprocess.CalledProcessError as error:
-            if error.returncode != PDCExitCodes.NO_FILES_FOUND:
-                raise error
-            LOG.error(f"No archived files found for PDC query: {search_pattern}")
-        LOG.info(f"Found archived files for PDC query: {search_pattern}")
+        for encryption_directory in [
+            self.encryption_directories.current,
+            self.encryption_directories.nas,
+            self.encryption_directories.pre_nas,
+        ]:
```
Comment (henrikstranneheim) [inline cg/meta/backup/backup.py:264]:
```suggestion
        for encryption_directory in  self.encryption_directories:
```
Would this work?

File: tests/meta/backup/conftest.py [hunk 3]
```diff
@@ -1,7 +1,28 @@
...
+    """Returns a mock method mimicking the pattern search made by the dsmc q archive command."""
+
+    def mock_method(search_pattern: str) -> list[str]:
+        match = fnmatch.filter(archived_flow_cells, search_pattern)
+        if not match:
+            raise subprocess.CalledProcessError(
+                cmd="dummy_method", returncode=PDCExitCodes.NO_FILES_FOUND
+            )
+        return match
+
```
Comment (henrikstranneheim) [inline tests/meta/backup/conftest.py:23]:
```suggestion
```

File: tests/meta/backup/conftest.py (hunk 1 above)
Comment (henrikstranneheim) [reply tests/meta/backup/conftest.py:24]: 👍

File: tests/meta/backup/conftest.py [hunk 4]
```diff
@@ -24,6 +45,16 @@ def dsmc_q_archive_output() -> list[str]:
     return output.splitlines()
 
 
+@pytest.fixture()
+def archived_flow_cells(encryption_directories: EncryptionDirectories) -> list[str]:
+    """Returns a list of archived flow cells."""
+    return [
+        f"{encryption_directories.current}/new_flow_cell.tar.gz.gpg",
```
Comment (henrikstranneheim) [inline tests/meta/backup/conftest.py:52]: Use FileExtension constants

File: tests/meta/backup/test_meta_backup.py [hunk 5]
```diff
@@ -15,32 +16,52 @@
...
 
-    # WHEN getting the dcms output of flow cell query
+    # WHEN querying pdc for a flow cell
     backup_api.query_pdc_for_flow_cell(flow_cell_id=flow_cell_name)
 
-    # THEN log that files were found
-    assert "Found archived files for PDC query: " in caplog.text
+    # THEN the flow cell is logged as found for one of the search patterns
+    assert fnmatch.filter(
+        names=caplog.messages, pat=f"Found archived files for PDC query:*{flow_cell_name}*.gpg"
```
Comment (henrikstranneheim) [inline tests/meta/backup/test_meta_backup.py:53]: Use FileExtension

**Review by henrikstranneheim — APPROVED** [review]
Great 👍

File: tests/meta/backup/conftest.py (hunk 3 above)
Comment (Vince-janv) [reply tests/meta/backup/conftest.py:23]: Black disagrees 🤔

Comment (henrikstranneheim) [reply tests/meta/backup/conftest.py:24]: ok, then we accept as is 👍

**Comment by Vince-janv** [comment]
### All three patterns are attempted:

```
11d [hiseq.clinical@hasta:~] [S_base] 20s $ cg -l DEBUG backup fetch-flow-cell -f HKHHGCCXX
Initialising Process with binary: /bin/dsmc
Use base call ['/bin/dsmc']
Initialising Process with binary: /bin/gpg
Use base call ['/bin/gpg']
Initialising Process with binary: /bin/tar
Use base call ['/bin/tar']
Instantiating status db
/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/store/api/base.py:37: SAWarning: relationship 'Collaboration.customers' will copy column collaboration.id to column customer_collaboration.collaboration_id, which conflicts with relationship(s): 'Customer.collaborations' (copies collaboration.id to customer_collaboration.collaboration_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="collaborations"' to the 'Collaboration.customers' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  return self.session.query(table)
/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/store/api/base.py:37: SAWarning: relationship 'Collaboration.customers' will copy column customer.id to column customer_collaboration.customer_id, which conflicts with relationship(s): 'Customer.collaborations' (copies customer.id to customer_collaboration.customer_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="collaborations"' to the 'Collaboration.customers' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  return self.session.query(table)
Processing flow cells: 0
HKHHGCCXX: retrieving from PDC
Starting DSMC command:
/bin/dsmc q archive /home/proj/stage/encrypt/*HKHHGCCXX*.gpg
Running command /bin/dsmc q archive /home/proj/stage/encrypt/*HKHHGCCXX*.gpg
Call ['/bin/dsmc', 'q', 'archive', '/home/proj/stage/encrypt/*HKHHGCCXX*.gpg'] exit with a non zero exit code
ANS1092W No files matching search criteria were found
No archived files found for PDC query: /home/proj/stage/encrypt/*HKHHGCCXX*.gpg
Starting DSMC command:
/bin/dsmc q archive /home/hiseq.clinical-stage/ENCRYPT/*HKHHGCCXX*.gpg
Running command /bin/dsmc q archive /home/hiseq.clinical-stage/ENCRYPT/*HKHHGCCXX*.gpg
Call ['/bin/dsmc', 'q', 'archive', '/home/hiseq.clinical-stage/ENCRYPT/*HKHHGCCXX*.gpg'] exit with a non zero exit code
ANS1092W No files matching search criteria were found
No archived files found for PDC query: /home/hiseq.clinical-stage/ENCRYPT/*HKHHGCCXX*.gpg
Starting DSMC command:
/bin/dsmc q archive /mnt/hds/proj/bioinfo-stage/TO_PDC/*HKHHGCCXX*.gpg
Running command /bin/dsmc q archive /mnt/hds/proj/bioinfo-stage/TO_PDC/*HKHHGCCXX*.gpg
Call ['/bin/dsmc', 'q', 'archive', '/mnt/hds/proj/bioinfo-stage/TO_PDC/*HKHHGCCXX*.gpg'] exit with a non zero exit code
ANS1092W No files matching search criteria were found
No archived files found for PDC query: /mnt/hds/proj/bioinfo-stage/TO_PDC/*HKHHGCCXX*.gpg
```

### With corrected paths it finds a really old flow cell
```
[hiseq.clinical@hasta:~] [S_base] 3s 8 $ /bin/dsmc q archive /mnt/hds/proj/bioinfo/TO_PDC/*HKHHGCCXX*.gpg
IBM Tivoli Storage Manager
Command Line Backup-Archive Client Interface
  Client Version 7, Release 1, Level 4.0
  Client date/time: 10/17/2023 16:54:16
(c) Copyright by IBM Corporation and other(s) 1990, 2015. All Rights Reserved.

Node Name: HASTA.SCILIFELAB.SE_CLINICAL
Session established with server BLACKHOLE: Linux/x86_64
  Server Version 8, Release 1, Level 9.300
  Server date/time: 10/17/2023 16:54:16  Last access: 10/17/2023 16:53:47

Accessing as node: SLLCLINICAL
             Size  Archive Date - Time    File - Expires on - Description
             ----  -------------------    -------------------------------
           607  B  04/10/2016 06:30:49    /mnt/hds/proj/bioinfo/TO_PDC/160202_ST-E00266_0064_AHKHHGCCXX.key.gpg Never Archive Date: 04/10/2016
   620,281,035 KB  04/10/2016 03:53:48    /mnt/hds/proj/bioinfo/TO_PDC/160202_ST-E00266_0064_AHKHHGCCXX.tar.gz.gpg Never Archive Date: 04/10/2016
```
### With corrected paths it finds a more recent NovaSeq 6000 flow cell
```
[hiseq.clinical@hasta:~] [S_base] 4s $ /bin/dsmc q archive /home/hiseq.clinical/ENCRYPT/*HKHMKDSX5*.gpg
IBM Tivoli Storage Manager
Command Line Backup-Archive Client Interface
  Client Version 7, Release 1, Level 4.0
  Client date/time: 10/17/2023 16:55:32
(c) Copyright by IBM Corporation and other(s) 1990, 2015. All Rights Reserved.

Node Name: HASTA.SCILIFELAB.SE_CLINICAL
Session established with server BLACKHOLE: Linux/x86_64
  Server Version 8, Release 1, Level 9.300
  Server date/time: 10/17/2023 16:55:33  Last access: 10/17/2023 16:54:16

Accessing as node: SLLCLINICAL
             Size  Archive Date - Time    File - Expires on - Description
             ----  -------------------    -------------------------------
 1,242,196,365 KB  03/14/2023 14:30:03    /home/hiseq.clinical/ENCRYPT/230310_A01901_0138_AHKHMKDSX5.tar.gz.gpg Never Archive Date: 03/14/2023
           607  B  03/14/2023 17:39:17    /home/hiseq.clinical/ENCRYPT/230310_A01901_0138_AHKHMKDSX5.key.gpg Never Archive Date: 03/14/2023
```

**Comment by Vince-janv** [comment]
Deployed successfully in production

**Pull request merged**